### PR TITLE
Fix resolution of excluded tasks for included builds

### LINKED ISSUE #24341: --exclude-task fails on complex included build structure starting with 7.6
We use a permanent included build that in turn includes a build for convention plugins. We apply a _script settings plugin_ to the included builds:

    // included_build/settings.gradle
    pluginManagement {
        includeBuild '../conventions'
    }

    plugins {
        id 'script-plugin'
    }

### Expected Behavior
Using `--exclude-task` in that setup should work normally, and up to and including 7.5.x, it does.

### Current Behavior
Starting from at least 7.6-rc-1 and also with 8.x, I immediately get the following error:
`The default project is not yet available for build.`

### Context
We're using that settings plugin to de-duplicate our complex plugin-repo declarations, so any other solution that accomplishes this would also work for us.

### Steps to Reproduce 
Run `gradlew build -x test` in this repo:
https://github.com/mkm-dt/gradle-included-build-settings-plugin-repro

### Your Environment
Should be irrelevant

### PR DESCRIPTION
**Pull request by king-tyler**
Fixes #24341

### Context
When a project with an included build, where the included build in turn includes a build for a convention plugin, Gradle will try to resolve excluded tasks before the main project is fully created when trying to configure convention plugin project.

This change fixes the behavior of the resolveExcludedTaskName method to return an empty filter if any of the build's projects have not yet created, the same that already occurs when the project is not yet loaded.

See also: #21936 where this bug was first introduced

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

**Comment by cobexer** [comment]
Sorry for the late reply.

Thank you for your contribution!

This PR has a valid DCO and tests. The relevant team for this area will confirm the design of the implementation choices.

**Comment by ov7a** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java [hunk 1]
```diff
@@ -61,6 +61,12 @@ public Filter resolveExcludedTaskName(BuildState defaultBuild, String taskName)
             // Too early to resolve excludes
             return new Filter(defaultBuild, Specs.satisfyNone());
         }
+        for (ProjectState project : defaultBuild.getProjects().getAllProjects()) {
```
Comment (abstratt) [inline subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java:64]: Could parallel `BuildState#isProjectsLoaded()` with a `BuildState.isProjectsCreated()`. 

Something like this:

```
@Override
    public boolean isProjectsCreated() {
        BuildProjectRegistry projectsForThisBuild = getProjectStateRegistry().findProjectsFor(getBuildIdentifier());
        return projectsForThisBuild != null && projectsForThisBuild.getAllProjects().stream().allMatch(ProjectState::isCreated);
    }
```

Didn't run any tests though.  Note that `getProjectStateRegistry().findProjectsFor(...)` returns `null` if project states are not available yet, whereas `getProjectStateRegistry().projectsFor(...)`  (called by`BuildState#getAllProjects()`) fails with an `IllegalStateException`.

Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java:64]: It seems that this new `isProjectsCreated` predicate could be used as the first guard in place of `isProjectsLoaded`, right?

Comment (abstratt) [reply subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java:64]: @bamboo points out that we should be able to check only the root project, as we do in the Configuratin Cache:

https://github.com/gradle/gradle/blob/93d70ab7fd76b71b116dc4237a15775f4cd5edac/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt#L857

That means we could have only:

```
     @Override
    public boolean isProjectsAvailable() {
        BuildProjectRegistry buildProjects = getProjectStateRegistry().findProjectsFor(getBuildIdentifier());
        return buildProjects != null && buildProjects.getRootProject().isCreated();
    }
```

and remove the extension method we have in the Configuration Cache.

Comment (abstratt) [reply subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java:64]: Also, we no longer need to check `isProjectsLoaded()` - as `isProjectsAvailable()` already does that check.

**Review by abstratt — CHANGES_REQUESTED** [review]
Thanks for your submission, @king-tyler. Some small changes requested. Please ask any questions if something is not clear.

**Comment by abstratt** [comment]
@king-tyler Sorry it took us so long to look into your submission. Are you still available and interested in working with us to get this PR ready for merge?

**Review by bamboo — APPROVED** [review]
Thanks for the patience!

**Pull request merged**